File: 250510/가운데에서 시작하여 빙빙 돌기/snail-start-from-center.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        // Please write your code here.
        int value = n*n;
        int[][] map = new int[n][n];
        // 동남서북 0123
        int[] dx = {0, 1, 0, -1};
        int[] dy = {1, 0, -1, 0};
        int directionIdx = 2;
        
        int x = n-1;
        int y = n-1+1;

        while (value >= 1) {
            int nx = x + dx[directionIdx];
            int ny = y + dy[directionIdx];

            if (nx<0 || nx>=n || ny<0 || ny>=n || map[nx][ny]!=0) {
                directionIdx = (directionIdx + 1) % 4;
            }
            
            x = x + dx[directionIdx];
            y = y + dy[directionIdx];

            map[x][y] = value;

            value--;
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                System.out.print(map[i][j] + " ");
            }
            System.out.println();
        }
    }
}


// (N - 1) / 2 => 가운데 인덱스
// 서 북 동 남
// 

File: 251021/가운데에서 시작하여 빙빙 돌기/snail-start-from-center.py
n = int(input())
grid = [[0] * n for _ in range(n)]

# Please write your code here.

curX = n - 1
curY = n - 1

# 동남서북
dx = [0,1,0,-1]
dy = [1,0,-1,0]

dir = 2

grid[curX][curY] = n*n

def inRange(x,y) :
    if x>=0 and x<n and y>=0 and y<n :
        return True
    return False

for i in range(n*n-1,0,-1) :
    nx = curX + dx[dir]
    ny = curY + dy[dir]

    if not inRange(nx,ny) or grid[nx][ny] != 0 :
        dir = (dir + 1) % 4
        nx = curX + dx[dir]
        ny = curY + dy[dir]

    grid[nx][ny] = i
    curX = nx
    curY = ny

for i in range(n) :
    for j in range(n) :
        print(grid[i][j],end=" ")
    print()